File: ui/src/pipeline/stages/scanning.ts
/*
 * Copyright 2026 OpenTrace Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Scanning stage: builds the structural graph (Repo, Dir, File, Package nodes)
 * and pre-computes lookup maps for downstream stages.
 */

import type {
  GraphNode,
  GraphRelationship,
  LoadingInput,
  ScanResult,
  PipelineContext,
  PipelineEvent,
  RepoFile,
} from '../types';
import {
  getExtension,
  detectLanguage,
  parentDir,
  ensureDirChain,
} from './loading';
import {
  isManifestFile,
  parseManifest,
  packageId,
  packageSourceUrl,
} from '../../runner/browser/parser/manifestParser';

const PARSEABLE_LANGUAGES = new Set([
  'python',
  'typescript',
  'javascript',
  'go',
  'rust',
  'java',
  'kotlin',
  'csharp',
  'c',
  'cpp',
  'ruby',
  'swift',
]);

export function* execute(
  input: LoadingInput,
  ctx: PipelineContext,
): Generator<PipelineEvent, ScanResult> {
  const { repo } = input;
  const repoId = `${repo.owner}/${repo.repo}`;

  yield {
    kind: 'stage_start',
    phase: 'scanning',
    message: `Scanning ${repo.files.length} files`,
  };

  const repoProps: Record<string, unknown> = { ref: repo.ref };
  if (repo.url) repoProps.source_uri = repo.url;
  if (repo.provider) repoProps.provider = repo.provider;

  const repoNode: GraphNode = {
    id: repoId,
    type: 'Repository',
    name: repoId,
    properties: repoProps,
  };

  const dirNodes = new Map<string, GraphNode>();
  const fileNodes: GraphNode[] = [];
  const structureRels: GraphRelationship[] = [];
  const parseableFiles: RepoFile[] = [];
  const packageNodes = new Map<string, GraphNode>();
  const dependencyRels: GraphRelationship[] = [];
  let goModulePath: string | undefined;

  // Pre-compute lookup maps for downstream stages
  const knownPaths = new Set<string>();
  const pathToFileId = new Map<string, string>();

  const total = repo.files.length;

  for (let i = 0; i < repo.files.length; i++) {
    if (ctx.cancelled) break;

    const file = repo.files[i];
    const ext = getExtension(file.path);
    const language = detectLanguage(ext);

    const fileId = `${repoId}/${file.path}`;
    const fileName = file.path.includes('/')
      ? file.path.slice(file.path.lastIndexOf('/') + 1)
      : file.path;

    const fileProps: Record<string, unknown> = {
      path: file.path,
      extension: ext,
    };
    if (language) fileProps.language = language;
    if (repo.url) {
      fileProps.source_uri = `${repo.url}/blob/${repo.ref}/${file.path}`;
    }

    fileNodes.push({
      id: fileId,
      type: 'File',
      name: fileName,
      properties: fileProps,
    });

    const dir = parentDir(file.path);
    ensureDirChain(repoId, dir, dirNodes, structureRels);

    const parentId = dir ? `${repoId}/${dir}` : repoId;
    structureRels.push({
      id: `${fileId}->DEFINED_IN->${parentId}`,
      type: 'DEFINED_IN',
      source_id: fileId,
      target_id: parentId,
    });

    if (language && PARSEABLE_LANGUAGES.has(language)) {
      parseableFiles.push(file);
    }

    // Populate lookup maps
    knownPaths.add(file.path);
    pathToFileId.set(file.path, fileId);

    // Manifest parsing: extract Package nodes + DEPENDS_ON rels
    if (isManifestFile(file.path)) {
      const manifestResult = parseManifest(file.path, file.content);

      if (file.path.endsWith('go.mod') || file.path.includes('/go.mod')) {
        const moduleMatch = file.content.match(/^module\s+(\S+)/m);
        if (moduleMatch) goModulePath = moduleMatch[1];
      }

      for (const dep of manifestResult.dependencies) {
        const pkgId = packageId(dep.registry, dep.name);
        if (!packageNodes.has(pkgId)) {
          const props: Record<string, unknown> = {
            registry: dep.registry,
            version: dep.version,
          };
          const sourceUrl = packageSourceUrl(dep.registry, dep.name);
          if (sourceUrl) props.source_uri = sourceUrl;

          packageNodes.set(pkgId, {
            id: pkgId,
            type: 'Package',
            name: dep.name,
            properties: props,
          });
        }

        dependencyRels.push({
          id: `${repoId}->DEPENDS_ON->${pkgId}`,
          type: 'DEPENDS_ON',
          source_id: repoId,
          target_id: pkgId,
          properties: {
            version: dep.version,
            dependency_type: dep.dependencyType,
            source: dep.source,
          },
        });
      }
    }

    yield {
      kind: 'stage_progress',
      phase: 'scanning',
      message: `Scanning ${file.path}`,
      detail: { current: i + 1, total, fileName: file.path },
    };
  }

  const structureNodes: GraphNode[] = [
    repoNode,
    ...dirNodes.values(),
    ...fileNodes,
    ...packageNodes.values(),
  ];

  yield {
    kind: 'stage_stop',
    phase: 'scanning',
    message: `Scanned ${fileNodes.length} files, ${dirNodes.size} directories, ${packageNodes.size} packages`,
    nodes: structureNodes,
    relationships: [...structureRels, ...dependencyRels],
  };

  return {
    repo,
    repoId,
    repoNode,
    dirNodes,
    fileNodes,
    structureRels,
    parseableFiles,
    packageNodes,
    dependencyRels,
    goModulePath,
    knownPaths,
    pathToFileId,
  };
}

File: agent/src/opentrace_agent/pipeline/scanning.py
# Copyright 2026 OpenTrace Contributors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Stage 1: Scanning — walk directory tree to produce structural nodes."""

from __future__ import annotations

import logging
from pathlib import Path
from typing import Generator

from opentrace_agent.pipeline.types import (
    EventKind,
    GraphNode,
    GraphRelationship,
    Phase,
    PipelineContext,
    PipelineEvent,
    PipelineInput,
    ProgressDetail,
    ScanResult,
    StageResult,
)
from opentrace_agent.sources.code.directory_walker import (
    DirectoryWalker,
)
from opentrace_agent.sources.code.import_analyzer import (
    package_id,
    package_source_url,
)
from opentrace_agent.sources.code.manifest_parser import (
    extract_go_module_path,
    parse_manifest,
)

logger = logging.getLogger(__name__)


def scanning(
    inp: PipelineInput,
    ctx: PipelineContext,
    out: StageResult[ScanResult],
) -> Generator[PipelineEvent, None, None]:
    """Walk directory tree and produce structural GraphNode/GraphRelationship objects."""
    yield PipelineEvent(
        kind=EventKind.STAGE_START,
        phase=Phase.SCANNING,
        message="Scanning directory tree",
    )

    root_path = Path(inp.path) if inp.path else None
    if root_path is None:
        yield PipelineEvent(
            kind=EventKind.ERROR,
            phase=Phase.SCANNING,
            message="No path provided",
            errors=["PipelineInput.path is required for local scanning"],
        )
        return

    root_path = root_path.resolve()
    repo_id = inp.repo_id or root_path.name
    repo_url = inp.repo_url
    ref = inp.ref
    provider = inp.provider

    if ctx.cancelled:
        return

    # Walk directory tree → flat nodes/rels
    walker = DirectoryWalker()
    walk = walker.walk(
        root_path=root_path,
        repo_id=repo_id,
        repo_name=root_path.name,
    )

    if ctx.cancelled:
        return

    nodes = walk.nodes
    rels = walk.relationships

    # Enrich Repository node with ref/sourceUri/provider
    for node in nodes:
        if node.type == "Repository":
            props = dict(node.properties or {})
            props["summary"] = f"Source code repository for {node.name}"
            if ref:
                props["ref"] = ref
            if repo_url:
                props["sourceUri"] = repo_url
            if provider:
                props["provider"] = provider
            nodes[nodes.index(node)] = GraphNode(
                id=node.id,
                type=node.type,
                name=node.name,
                properties=props,
            )
            break

    # Enrich File nodes with sourceUri
    if repo_url and ref:
        for i, node in enumerate(nodes):
            if node.type == "File":
                path = (node.properties or {}).get("path")
                if path:
                    props = dict(node.properties or {})
                    props["sourceUri"] = f"{repo_url}/blob/{ref}/{path}"
                    nodes[i] = GraphNode(
                        id=node.id,
                        type=node.type,
                        name=node.name,
                        properties=props,
                    )

    # Parse manifest files → Package nodes + DEPENDS_ON rels
    go_module_path: str | None = None
    package_nodes: dict[str, GraphNode] = {}
    seen_dep_rels: set[str] = set()

    for manifest_path, manifest_abs_path in walk.manifest_files:
        try:
            content = Path(manifest_abs_path).read_text(encoding="utf-8", errors="replace")
        except (OSError, IOError) as exc:
            logger.warning("Could not read manifest %s: %s", manifest_abs_path, exc)
            continue

        # Extract go module path from go.mod
        basename = manifest_path.rsplit("/", 1)[-1]
        if basename == "go.mod":
            go_module_path = extract_go_module_path(content)

        parsed_deps = parse_manifest(manifest_path, content)
        for dep in parsed_deps:
            pkg_id = package_id(dep.registry, dep.name)

            # Create Package node (deduplicated)
            if pkg_id not in package_nodes:
                pkg_props: dict[str, str] = {"registry": dep.registry}
                source_url = package_source_url(dep.registry, dep.name)
                if source_url:
                    pkg_props["sourceUri"] = source_url
                package_nodes[pkg_id] = GraphNode(
                    id=pkg_id,
                    type="Dependency",
                    name=dep.name,
                    properties=pkg_props,
                )

            # DEPENDS_ON relationship (deduplicated by rel ID)
            rel_id = f"{repo_id}->DEPENDS_ON->{pkg_id}"
            if rel_id not in seen_dep_rels:
                seen_dep_rels.add(rel_id)
                rels.append(
                    GraphRelationship(
                        id=rel_id,
                        type="DEPENDS_ON",
                        source_id=repo_id,
                        target_id=pkg_id,
                        properties={
                            "version": dep.version,
                            "dependency_type": dep.dependency_type,
                            "source": dep.source,
                        },
                    )
                )

    # Add package nodes to the structural nodes list
    nodes.extend(package_nodes.values())

    yield PipelineEvent(
        kind=EventKind.STAGE_STOP,
        phase=Phase.SCANNING,
        message=f"Scanned {len(nodes)} nodes, {len(walk.file_entries)} parseable files",
        detail=ProgressDetail(current=len(nodes), total=len(nodes)),
        nodes=nodes,
        relationships=rels,
    )

    out.value = ScanResult(
        repo_id=repo_id,
        root_path=str(root_path),
        structural_nodes=nodes,
        structural_relationships=rels,
        file_entries=walk.file_entries,
        known_paths=walk.known_paths,
        path_to_file_id=walk.path_to_file_id,
        go_module_path=go_module_path,
        repo_url=repo_url,
        ref=ref,
        provider=provider,
    )
